assets/libraries/common.rego

```diff
@@ -433,3 +433,242 @@ any_principal(statement) {
+    regex_compatible_principals = [{lower(key): convert_value_to_regex(principal, key)} | key := available_keys[_]]
+} else = regex_compatible_principals{
+	principal := object.get(statement, principal_field, [])
+    principal == []
+    regex_compatible_principals = []
+}
+
+# Returns normalized regex compatible policy statement
+normalize_statement(statement) = ps {
+	ps = {
+         "principal": handle_principle_regex(statement, "Principal"),
+         "not_principal": handle_principle_regex(statement, "NotPrincipal"),
+         "effect": statement.Effect,
+         "action": convert_value_to_regex(statement, "Action"),
+         "not_action": convert_value_to_regex(statement, "NotAction"),
+         "resource": convert_value_to_regex(statement, "Resource"),
+         "not_resource": convert_value_to_regex(statement, "NotResource"),
+         # Condition is not normalized as other fields since it is technically not feasible to 
+         # evaluate in a declarative fashion. The consumer may have to manually evaluate the 
+         # condition based on the context of a control
+         "condition": object.get(statement, "Condition", []),
+         "sid": object.get(statement, "Sid", "")
+          }
+}
+
+# Parses policy statements to regex based policy
+make_regex_compatible_policy_statement(policy_statements) = regex_compatible_statements {
+	regex_compatible_statements := [normalize_statement(statement) | statement := normalize_to_list(policy_statements)[_]]
+}
+
+# This method helps in evaluating if an action matches in a statement.
+# In case true is returned , it just means that this action is affected by this statement. 
+# One has to explicitly check whether it is allowed / denied
+statement_matches_action(statement, action) {
+	object.get(statement, "not_action", []) != []
+    not has_atleast_one_match(action, object.get(statement, "not_action", []))
+} else {
+	object.get(statement, "action", []) != []
+    has_atleast_one_match(action, object.get(statement, "action", []))
+}
+
+# This method helps in evaluating if a resource matches in a statement.
+# In case true is returned , it just means that this resource's access is affected by this statement. 
+# One has to explicitly check whether it is allowed / denied.
+statement_matches_resource(statement, resource) {
+	object.get(statement, "not_resource", []) != []
+    not has_atleast_one_match(resource, object.get(statement, "not_resource", []))
+} else {
+	object.get(statement, "resource", []) != []
+    has_atleast_one_match(resource, object.get(statement, "resource", []))
+}
+
+# This method helps in evaluating if a principal matches in a statement.
+# In case true is returned , it just means that this principal's access is affected by this statement. 
+# One has to explicitly check whether it is allowed / denied.
+statement_matches_principal(statement, principal, principal_type) {
```

Comment:
I think Principal can be also set as a string

Reply:
Hey yes you are right. But it has been handled in `make_regex_compatible_policy_statement` method. So all statements that are consumed by the functions implemented in this PR requires the policy to be normalized by the `make_regex_compatible_policy_statement`. The values irrespective of being a list or string for Principal / Resource / Actions are normalized to a list by default for easy evaluation later.  You can check that in line 475.

Let me know if any changes are to be done here.

Reply:
You are right! My bad, sorry!

Can you fix the merge conflicts, please?